### PR TITLE
Victor yang chlg

### PR DESCRIPTION
**Pull request by yanky2000**
Title: Adds display mode for peers
TLDR;
Adds button in the left menu to start controlled "Display mode" for peers - aka "Property demo" by location.
During the demo all users can browse through available pictures of the same property individually, but the property itself can be changed only by Initiator.

Description:
Two proxy API routes were added to store API keys & endpoint securely on the server.
The aggregated data (property list, adjusted according to a user-specified location) is calculated on the initiator's client side and stored in indexdb.
When the carousel is initiated, it signals its "actions" to peer users via news events. The first "start" event goes with a payload of an aggregated property list, through which everybody will browse afterwards.

deps:

splidejs - carousel library
idbkeyval - to interact with indexDb
** Add keys & endpoint to .env file to run.

I tried to stay consistent with the techniques used in the rest of the codebase, however there are some key points to consider:

introducing typescript could be very beneficial for dx and overall type safety
a build system (with auto-reload of assets) and use of modules would be very helpful
extensive use of global scope may pose a problem in some cases
direct dom & styles manipulation may hurt perf
some tests would be appreciated

### CODE REVIEW

File: public/js/client.js [hunk 1]
```diff
@@ -9802,3 +9954,167 @@ function sanitizeXSS(src) {
...
+
+// Carousel helper functions ----[START]----->
+async function fetchCarouselData(targetAddressString) {
+    // TODO: handle network errors more consistently
+    const targetLocationPromise = await fetchLocationCoordinates(targetAddressString);
+    const availablePropertiesLocationsPromise = await fetchProperties();
+
+    try {
+        const values = await Promise.all([targetLocationPromise, availablePropertiesLocationsPromise]);
+        const [targetLocationCoords, propertiesLocations] = values;
```
Comment (ofekrom) [inline public/js/client.js:10060]: redundant line

File: public/js/client.js [hunk 2]
```diff
@@ -9802,3 +9954,167 @@ function sanitizeXSS(src) {
...
+// <---[END]----- Carousel management
+
+// Carousel helper functions ----[START]----->
+async function fetchCarouselData(targetAddressString) {
+    // TODO: handle network errors more consistently
+    const targetLocationPromise = await fetchLocationCoordinates(targetAddressString);
+    const availablePropertiesLocationsPromise = await fetchProperties();
+
+    try {
+        const values = await Promise.all([targetLocationPromise, availablePropertiesLocationsPromise]);
```
Comment (ofekrom) [inline public/js/client.js:10059]: This is not needed

File: public/js/client.js [hunk 3]
```diff
@@ -9802,3 +9954,167 @@ function sanitizeXSS(src) {
...
+async function updateCarousel(currentImageGroupIdx, currentImageItemIdx) {
+    const newPropertyImageList = await getObjectImageLinksFromDb(currentImageGroupIdx, currentImageItemIdx);
+    addReplaceImageLinksInCarousel(newPropertyImageList);
+}
+// <---[END]----- Carousel management
+
+// Carousel helper functions ----[START]----->
+async function fetchCarouselData(targetAddressString) {
+    // TODO: handle network errors more consistently
+    const targetLocationPromise = await fetchLocationCoordinates(targetAddressString);
```
Comment (ofekrom) [inline public/js/client.js:10055]: This could have been a promise.all

File: public/js/client.js [hunk 4]
```diff
@@ -9802,3 +9954,167 @@ function sanitizeXSS(src) {
...
+async function storeImagesInDb(propertyList) {
+    try {
+        await idbKeyval.set(PROPERTY_LIST, propertyList);
+    } catch (error) {
+        console.log(error);
+    }
+}
+
+async function getObjectImageLinksFromDb(groupIdx, itemIdx) {
+    let propList;
```
Comment (ofekrom) [inline public/js/client.js:10082]: I didn't like the implementation of this function.
the return should be inside the try-catch

File: public/js/client.js [hunk 5]
```diff
@@ -9802,3 +9954,167 @@ function sanitizeXSS(src) {
...
+    }
+    return propList[groupIdx]?.[itemIdx]?.images;
+}
+
+function addReplaceImageLinksInCarousel(newImageLinks) {
+    const splideList = document.querySelector('.splide__list');
+    splideList.innerHTML = ''; // This clears out all existing slides
+
+    newImageLinks.forEach((url) => {
+        var newSlide = document.createElement('li');
```
Comment (ofekrom) [inline public/js/client.js:10096]: var?

File: public/js/client.js [hunk 6]
```diff
@@ -9802,3 +9954,167 @@ function sanitizeXSS(src) {
...
+
+function hideCarouselControls() {
+    Array.from(controlBtns).forEach((el) => el.classList.toggle('hidden', true));
+}
+
+
+async function initCarousel({ propertyList, hasControls = false }) {
+    elemDisplay(myVideoAvatarImage, false);
+    elemDisplay(myVideoWrap, false);
+    toggleVideoPin(pinVideoPositionSelect.value, true);
```
Comment (ofekrom) [inline public/js/client.js:10005]: It would be better to use a different function with shared functionality. The force flag is too hacky IMO.

File: public/js/client.js [hunk 7]
```diff
@@ -7591,6 +7718,7 @@ async function emitPeersAction(peerAction) {
         peer_use_video: useVideo,
         peer_action: peerAction,
         send_to_all: true,
+        ...(payload ? { payload: payload } : {}),
```
Comment (ofekrom) [inline public/js/client.js:7721]: ...(payload && { payload }),

File: public/js/client.js [hunk 8]
```diff
@@ -9802,3 +9954,167 @@ function sanitizeXSS(src) {
...
+}
+
+function addReplaceImageLinksInCarousel(newImageLinks) {
+    const splideList = document.querySelector('.splide__list');
+    splideList.innerHTML = ''; // This clears out all existing slides
+
+    newImageLinks.forEach((url) => {
+        var newSlide = document.createElement('li');
+        newSlide.className = 'splide__slide';
+        newSlide.innerHTML = `<img src="${url}" alt="">`;
```
Comment (ofekrom) [inline public/js/client.js:10098]: inline HTML is bad practice

File: public/js/client.js [hunk 9]
```diff
@@ -9802,3 +9954,167 @@ function sanitizeXSS(src) {
 function disable(elem, disabled) {
     elem.disabled = disabled;
 }
+
+//  Carousel control button click handlers  ---[START]------->
+async function handleDisplayModePrevItem() {
+    if (currentItemIdx === 0) {
```
Comment (ofekrom) [inline public/js/client.js:9960]: This code is hard to understand. I suggest using an external class for managing the carousel items

File: public/js/client.js [hunk 10]
```diff
@@ -9802,3 +9954,167 @@ function sanitizeXSS(src) {
...
+// Carousel helper functions ----[START]----->
+async function fetchCarouselData(targetAddressString) {
+    // TODO: handle network errors more consistently
+    const targetLocationPromise = await fetchLocationCoordinates(targetAddressString);
+    const availablePropertiesLocationsPromise = await fetchProperties();
+
+    try {
+        const values = await Promise.all([targetLocationPromise, availablePropertiesLocationsPromise]);
+        const [targetLocationCoords, propertiesLocations] = values;
+        const sortedPropertyList = getSortedLocationChunks(propertiesLocations, targetLocationCoords);
```
Comment (ofekrom) [inline public/js/client.js:10061]: sorting should have been done on the backend.

File: public/js/client.js [hunk 11]
```diff
@@ -9802,3 +9954,167 @@ function sanitizeXSS(src) {
 function disable(elem, disabled) {
     elem.disabled = disabled;
 }
+
+//  Carousel control button click handlers  ---[START]------->
+async function handleDisplayModePrevItem() {
+    if (currentItemIdx === 0) {
+        const propList = await idbKeyval.get(PROPERTY_LIST);
```
Comment (ofekrom) [inline public/js/client.js:9961]: It's hard to understand where idbKeyval comes from, I had to invest a few minutes of research to find that it comes from a library.

File: public/js/locationHelpers.js [hunk 12]
```diff
@@ -0,0 +1,44 @@
+// Haversine formula to calculate distance between two points
+function getDistance(lat1, lon1, lat2, lon2) {
```
Comment (ofekrom) [inline public/js/locationHelpers.js:2]: Good job using a separate file for these functions
